crates/ide-completion/src/render/function.rs

```diff
@@ -157,6 +169,65 @@ fn render(
     item
 }
 
+fn compute_function_match(
+    ctx: &RenderContext<'_>,
+    func: hir::Function,
+    func_kind: &FuncKind<'_>,
+    db: &dyn HirDatabase,
+) -> Option<CompletionRelevanceFn> {
+    if func.as_assoc_item(ctx.db()).is_none() {
+        return None;
+    }
+
+    let has_args = func.num_params(db) > 0;
+    let ret_type = func.ret_type(db);
+    let has_self_arg = func.self_param(db).is_some();
+
+    let self_type = match func_kind {
+        FuncKind::Function(PathCompletionCtx {
+            qualified: Qualified::With { resolution, .. },
+            ..
+        }) => {
+            let Some(hir::PathResolution::Def(hir::ModuleDef::Adt(hir::Adt::Struct(struct_)))) =
+                resolution
+            else {
+                return None;
+            };
+
+            Some(struct_.ty(db))
+        }
+        _ => None,
+    };
```

Comment:
This is the wrong way to fetch the self type for the impl. For one because it only works for structs, and the other reason being that this discards any generics of the struct. Instead we can make use of the `as_assoc_item` result as that gives us an `AssocItem`,  then you can fetch the self type via `AssocItem::implementing_ty` (its a new function, so you might need to rebase on latest master)